### ISSUE TITLE
special characters in Coldfusion reports

### ISSUE THREAD

**Issue by ekrimmel** [ArctosDB/arctos#1261]
Arctos does not like my report when the data going into the report has a quotation mark, e.g. I get the error message below when trying to print the report *CHAS_label_vert* for http://arctos.database.museum/guid/CHAS:Egg:569 because of the quote mark in the Specific Locality ("Warren Township, west side of Gage's Lake"). Any suggestions on how to fix the problem aside from deleting the quote from the data?

![capture](https://user-images.githubusercontent.com/11165319/30185655-125c4a96-93e8-11e7-9893-4fbf9d411208.JPG)

**Comment by dustymc** [comment]
That's not a quote, it's a HTML encoded character.

```
UAM@ARCTOS> select spec_locality from flat where guid='CHAS:Egg:569';

SPEC_LOCALITY
------------------------------------------------------------------------------------------------------------------------
Warren Township, west side of Gage&#39;s Lake


```

The hash in CF means "this is the start of a variable." I don't think it's possible to avoid processing that in IIF().

Maybe just because of my language preferences, I'd assemble the weird locality string in the DB

```
select
decode(replace(higher_geog,'North America, United States','USA'),
'no higher geography recorded','',
higher_geog
)
||
decode(
replace(spec_locality,'no specific locality recorded'),
NULL,NULL,
', ' || spec_locality
) locstring
 12  from flat where guid='CHAS:Egg:569';

LOCSTRING
------------------------------------------------------------------------------------------------------------------------
North America, United States, Illinois, Lake County, Warren Township, west side of Gage&#39;s Lake
```

which should let you avoid IIF, and hopefully this problem, altogether.

**Comment by ekrimmel** [comment]
@dustymc thank you! I like your way of assembling the locality string in the database much better. I'm still very much in the "trying to get a handle on using/writing reports for Arctos" phase...

**Comment by dustymc** [comment]
Here's a better version - the original can leave some weird stuff in some situations.

```
select distinct
	trim(trim(leading ',' from 
		decode(higher_geog,
			'no higher geography recorded',NULL,
			replace(higher_geog,'North America, United States','USA')
		)
		||
		decode(
			replace(spec_locality,'no specific locality recorded'),
			NULL,'',
			', ' || spec_locality
		)
	)) locstring
from flat where guid='CHAS:Egg:569';
```

An alternative is a DB function, which you would call with 
```
select...doWeirdThing(someIdOrStringOrWhatever) AS someLocalVariable.... 
```
Those are generally easy to write, can do stuff that's difficult in straight SQL (concatenate rows into a string, etc.), and are available for anyone. There are a bunch of them at https://github.com/ArctosDB/DDL/tree/master/functions.

I like https://github.com/ChicagoAcademyofSciences/arctos-reports - you should write up a brief how-to.

I think we're all perpetually in the "trying to get a handle on using/writing reports for Arctos" phase! Doing stuff in the DB should also make your reports more future-proof; the data should work in any report generator.

**Comment by ekrimmel** [comment]
Oh man, I dipped my head into https://github.com/ArctosDB/DDL/tree/master/functions briefly a while back and got confused quickly. Hard to follow without documentation, and with variables named inconsistently across functions. Which totally makes sense for those scripts being written on an as-needed basis by different people for different institutions, but maybe sometime in the future merits someone going back and cleaning up/writing documentation.

For writing a DB function, do those then need to get called with a pre-function? And how does that work? What I think I vaguely understand from scrounging around is that...
1. pre-functions for reports exist in *[label_functions.cfm](https://github.com/ArctosDB/arctos/blob/ac94b4ed81db900bfe0b1bd62531172781954e77/Reports/functions/label_functions.cfm)*
1. from *label_functions.cfm* another, stand-alone function is called, e.g. *[getLoan.cfm](https://github.com/ArctosDB/arctos/blob/3fb05cef0d6cbe67579b3e2bdcb96ce1a0549f6d/picks/getLoan.cfm)*
1. that function feeds data back to the SQL in the report handler
1. the report handler feeds data into the ColdFusion Report Builder file, ideally just for formatting

Also yea, I'd like to use my learning curve to write up a brief how-to but feel like I need to be a little further along the in getting of the handle... if I wanted to get more familiar with DDL do you have any resources to suggest? I'm used to more basic SQL stuff on the DML side.

**Comment by dustymc** [comment]
I mostly like the organizational idea of a report repository - whatever's in it! That seems like something all collections/institutions would benefit from.

Here's more on the labels themselves. 

Pre-functions are ColdFusion - they can do a bit more than even Oracle functions (eg, merge records). (That can be done in Oracle Procedures, but those are less accessible.) Pre-functions DO have weird variable names and such, because they're mostly aimed at one collection doing one weird thing. Maybe I misjudge that and they should be made more portable?? I did add some documentation recently - https://github.com/ArctosDB/arctos/blob/master/Reports/functions/label_functions.cfm.

Oracle functions - https://github.com/ArctosDB/DDL/tree/master/functions - are a different thing entirely. You can call them directly in the SQL. You name your variables, "my" (internal) variables are anonymous from your perspective. To recycle my example above, all you need to know is the function name ("doWeirdThing") and the number/order/datatype of inputs ("someIdOrStringOrWhatever"). You don't have to name someIdOrStringOrWhatever anything, just send data. "Select doWeirdThing(flat.appropriate_data) as myVariableName from flat where guid='Bla:Blah:6'." Naming the result happens in the SQL - "AS myVariableName". Oracle functions all work the same - some (https://docs.oracle.com/cd/B19306_01/server.102/b14200/functions001.htm) come built-in, others we've developed. I'll try to review the documentation, although I think it's pretty decent for most anything that would be useful in a report.

I also added a paragraph to http://handbook.arctosdb.org/how_to/How-to-Create-Labels.html.

> getLoan

You (or GitHub automagically??) linked the popup from which you select a loan (eg, to add to project).

Pre-function get_loan is calling another CF object, [CustomTags/getLoanFormInfo.cfm](https://github.com/ArctosDB/arctos/blob/master/CustomTags/getLoanFormInfo.cfm), which isn't exactly documented but I think is clear - please let me know what I can do to make it more understandable.

The data-->report path in the reporter is:

1. SQL, including Oracle functions
2. Optionally pre-functions, which may leverage (and are essentially an abstract of) Oracle functions, CF functions, CustomTags, Java, remote webservices, or anything else I can get to from ColdFusion.
3. Back to the reporter, which is just a convenient wrapper for handling all this
4. The CF report itself

Everything you do using (1) is extremely portable. If you decide you like CrystalReports/tex/Word/whatever better than the CF reporter, just take your SQL and use it. If you've piped it through the pre-function we'd have to add a code blob to get it out in the format you need. If you're manipulating it (eg, IIF() statements) in the report builder, you'll have to rewrite that code in your new tool.
